[Gokul Rajiv] iP

gok99 opened the pull request:
### Duke
> "Duke rhymes with fluke" - Margaret Thatcher ([source](https://www.youtube.com/watch?v=cvh0nX08nRw))

Duke helps you keep track of your tasks and todos. It does the following:
- Barely works
- 

To get started with Duke:
1. Download Duke from [here](https://github.com/gok99/ip/releases/tag/v0.1)
2. ~~Wrestle issues with your `jdk`~~
3. Write your own version of Duke that actually does what you want

Features:
- [x] Bugs
- [ ] Be useful (in progress)

---

Here's a code snippet that does something:
```java
public class Main {
    public static void main(String[] args) {
        main(args);
    }
}
```

Commit:
Add Gradle support

gok99 committed:
fix: Ignore .class files and fix ACTUAL.TXT path

gok99 committed:
add: Implement greet, echo and exit

gok99 committed:
add: Implement add and list

gok99 committed:
add: Implement done marking functionality

gok99 committed:
add: Implement ToDos, Events and Deadlines

gok99 committed:
fix: Make runtest.sh executable

gok99 committed:
fix: Use scanner in InputManager

gok99 committed:
add: Automated testing

gok99 committed:
add: Error handling

gok99 committed:
add: Delete feature

gok99 committed:
fix: Update tests

gok99 committed:
add: Remaining documentation

gok99 committed:
refactor: Organise files into packages

gok99 committed:
Add saving feature

gok99 committed:
Add extended datetime support and formatting

gok99 committed:
Merge branch 'branch-level-7'

gok99 committed:
Merge branch 'branch-level-8'

gok99 committed:
Refactor classes and employ more OOP

gok99 committed:
Add JUnit tests

gok99 committed:
Fix indentation to 4 spaces

gok99 committed:
Update documentation

gok99 committed:
Add find feature

gok99 committed:
Fix formatting issues

gok99 committed:
Merge branch 'branch-A-JavaDoc' into master

gok99 committed:
Resolve conflicts and merge branch-A-CodingStandard to master

gok99 committed:
Resolve conflicts and merge branch-Level-9 into master

gok99 committed:
Fix unmerged branch-A-CodingStandard branch to master

gok99 committed:
Merge branch 'add-gradle-support'

gok99 committed:
Fix gradle mainClassName

gok99 committed:
Add checkstyle and fix some style issues

gok99 committed:
Add initial GUI

gok99 committed:
Merge branch 'branch-Level-10'

zacharylwy commented on src/main/java/bot/TaskList.java:
```diff
@@ -0,0 +1,93 @@
+        }
+        Task task = getTaskAt(index);
+        this.taskList.remove(index);
+        return new String[]{
```
{ should have space before

zacharylwy commented on src/main/java/commands/DoneCommand.java:
```diff
@@ -0,0 +1,27 @@
+        }
+        Task task = bot.getTaskList().getTaskAt(index);
+        task.markDone();
+        return new String[]{
```
{ should have space before

zacharylwy reviewed (commented):
Very consistent work and style. Keep up the good work

sj1999-BIT commented on src/main/java/commands/EventCommand.java:
```diff
@@ -14,7 +14,7 @@
+    public String[] run(Bot bot, String[] args) {
         if (args[0].equals("")) {
             throw new InvalidArgumentsException(Ui.ERROR_SIGNATURE + "The description of an event cannot be empty.");
         }
```
This two seems like mutually exclusive events, perhaps you could use if else to connect them? 
For example, you could create the splitArgs array at the start. Then if splitArgs[0].equals(""), give first message, else if splitArgs.length < 2, give second message. 

I personally recommend avoid single if statements recurring in one function, as you may missed out certain logically flaws that the current if condition could not catch, but could lead to runtime exceptions.

gok99 replied:
Point well taken! In this case, I don't think there are any functional differences but combining the two definitely makes the mutual exclusivity more obvious.

sj1999-BIT commented on src/main/java/bot/Storage.java:
```diff
@@ -0,0 +1,76 @@
+
+    public static final String SAVE_FILE_PATH = "data/save.txt";
+
+    private static void write(List<String> texts) {
```
I would recommend a comment above this function as its seems to be quite a key function for this class. Perhaps just a simple description? I had a similar logic in my own work and I had to update my write-alike function several times as I keep updating the input format.

So perhaps it would be better to have a function descriptor for it?

sj1999-BIT commented on src/main/java/bot/TaskList.java:
```diff
@@ -0,0 +1,93 @@
+    /**
+     * Set the taskList
+     *
+     * @param taskList list of tasks
```
I think that it would be better to avoid the setters as it would reduce the point of having a private tasklist. 
I personally prefer to only allow the private attributes be modified only via the public methods provided by the wrapper class. For example in this case I only allow the tasklist to be modified by the add and deleteTask functions, whereas the getter will create a new arraylist and copy the private tasklist over, so as to return this copy instead of the actual tasklist. 
This is prevent others from modifying the tasklist at will, which could lead to runtime exceptions

gok99 replied:
Also a good point, was definitely being lazy with setters in various places. Some 'tell-don't-ask' seems due!

sj1999-BIT commented on src/main/java/bot/TaskType.java:
```diff
@@ -0,0 +1,25 @@
+public enum TaskType {
+
+    Deadline("D"), Event("E"), Todo("T"), Blank("B");
+
```
This is quite cool. I did not know how to exactly use the functions in enum, so my work just used them like a pointer in C class.

gok99 committed:
Add assertions where inputs were assumed to be non-null

Previously, some functions made assumptions about the input - that they will always be non-null.

Add assertions in some of these functions that the input not be null, so that errors can be narrowed down to these sections of code.

gok99 committed:
Improve code quality with SLAP, etc

Bits of code quality in areas of the codebase could have been improved

Implement SLAP and guard clause improvements

gok99 committed:
Merge pull request #1 from gok99/branch-A-Assertions

Add assertions

gok99 committed:
Resolve conflicts and merge master

gok99 committed:
Merge pull request #2 from gok99/branch-A-CodeQuality

Improve code quality

gok99 committed:
Add MassOps for delete

gok99 committed:
Add MassOps for done

gok99 committed:
Add CI

gok99 committed:
Temporarily remove some checkstyle asserts

gok99 committed:
Update cli tests

gok99 committed:
Add initial User Guide description

gok99 committed:
Add missing hyphen to User Guide

gok99 committed:
Add User Guide features and usage

gok99 committed:
Merge branch 'master' of https://github.com/gok99/ip

gok99 committed:
Remove IO redirection in workflow

gok99 committed:
Improve GUI and add temmie theme

gok99 committed:
Fix save file creation on init, minor doc fixes

gok99 committed:
Add Ui.png

gok99 committed:
Fix javafx dependencies for cross platform

gok99 committed:
Fix scrollbar position

gok99 committed:
Fix build.gradle further

gok99 committed:
Update Ui.png

gok99 committed:
Fix some header comments

gok99 committed:
Merge pull request #4 from gok99/fixStyle

Fix some header comments
